File: Queue.cpp
#include "Queue.h"

bool Queue::is_empty() const{
    return head == nullptr;
}

void Queue::push(string value){
    ONode* new_elem = new ONode(value);

    if (is_empty()){
        head = new_elem;
        tail = new_elem;
        return;
    }

    tail->next = new_elem;
    tail = new_elem;
}

void Queue::pop(){
    if (is_empty()){
        throw out_of_range("Queue is empty");
    }

    ONode* new_head = head->next;
    cout << "Popped value: " << head->data << endl;
    delete head;
    head = new_head;
}

string Queue::get_head() const
{
    return head->data;
}

void Queue::print() const{
    ONode* curr = head;

    while (curr->next != nullptr){
        cout << curr->data << "<-";
        curr = curr->next;
    }
    cout << curr->data << endl;
}

File: go_lb3/queue.go
package main

import (
	"fmt"
	"strings"
)

type Node struct {
	Data string
	Next *Node
}

type Queue struct {
	Head *Node
}

func construct_queue() *Queue {
	return &Queue{}
}

func (q *Queue) push(element string) {
	newNode := &Node{Data: element}

	if q.Head == nil {
		q.Head = newNode
		return
	}

	current := q.Head
	for current.Next != nil {
		current = current.Next
	}
	current.Next = newNode
}

func (q *Queue) pop() (string, error) {
	if q.Head == nil {
		return "", fmt.Errorf("Queue is empty")
	}

	popped := q.Head
	q.Head = q.Head.Next
	return popped.Data, nil
}

func (q *Queue) print() string {
	var sb strings.Builder
	sb.WriteString("Queue: [")
	curr := q.Head
	for curr != nil {
		if curr.Next != nil {
			sb.WriteString(fmt.Sprintf("%q", curr.Data))
			sb.WriteString(", ")
		} else {
			sb.WriteString(fmt.Sprintf("%q", curr.Data))
		}
		curr = curr.Next
	}
	sb.WriteString("]")
	return sb.String()
}
